Tools/scripts/freeze_modules.py

```diff
@@ -113,67 +169,168 @@ def parse_frozen_spec(rawspec, knownids=None, section=None):
+        )
 
-def _parse_frozen_specs(rawspecs, section, seen):
-    for rawspec in rawspecs:
-        spec, subs = parse_frozen_spec(rawspec, seen, section)
-        yield spec
-        for spec in subs or ():
-            yield spec
+    def get_array_size(self):
+        with open(self.frozenfile) as frozenfile:
+            text = frozenfile.read()
+        return text.count(',')
 
 
-def resolve_frozen_file(spec, destdir=MODULES_DIR):
-    if isinstance(spec, str):
-        modname = spec
-    else:
-        _, frozenid, _, _, _= spec
-        modname = frozenid
+def resolve_frozen_file(frozenid, destdir=MODULES_DIR):
+    """Return the filename corresponding to the given frozen ID.
+
+    For stdlib modules the ID will always be the full name
+    of the source module.
+    """
+    if not isinstance(frozenid, str):
+        try:
+            frozenid = frozenid.frozenid
+        except AttributeError:
+            raise ValueError(f'unsupported frozenid {frozenid!r}')
     # We use a consistent naming convention for all frozen modules.
-    return os.path.join(destdir, modname.replace('.', '_')) + '.h'
+    frozenfile = frozenid.replace('.', '_') + '.h'
+    if not destdir:
+        return frozenfile
+    return os.path.join(destdir, frozenfile)
 
 
-def resolve_frozen_files(specs, destdir=MODULES_DIR):
+def resolve_frozen_files(modules, destdir=MODULES_DIR):
     frozen = {}
     frozenids = []
     lastsection = None
-    for spec in specs:
-        frozenid, pyfile, *_, section = spec
-        if frozenid in frozen:
+    for mod in modules:
+        src, section = FrozenSource.from_raw(mod, destdir)
+        if src.frozenid in frozen:
             if section is None:
                 lastsection = None
             else:
                 assert section == lastsection
             continue
         lastsection = section
-        frozenfile = resolve_frozen_file(frozenid, destdir)
-        frozen[frozenid] = (pyfile, frozenfile)
+        frozen[frozenid] = src
         frozenids.append(frozenid)
     return frozen, frozenids
 
 
+#######################################
+# frozen modules
+
+class FrozenModule(namedtuple('FrozenModule', 'name ispkg section source')):
+
+    def __getattr__(self, name):
+        return getattr(self.source, name)
+
+    @property
+    def modname(self):
+        return self.name
+
+    def summarize(self):
+        source = self.source.modname
+        if source:
+            source = f'<{source}>'
+        else:
+            source = os.path.relpath(self.pyfile, ROOT_DIR)
+        return {
+            'module': self.name,
+            'ispkg': self.ispkg,
+            'source': source,
+            'frozen': os.path.basename(self.frozenfile),
+            'checksum': _get_checksum(self.frozenfile),
+        }
+
+
+def _iter_sources(modules, destdir=MODULES_DIR):
+    seen = set()
+    for mod in modules:
+        if mod.source not in seen:
+            yield mod.source
+            seen.add(mod.source)
```

Comment:
This doesn't seem to use the destdir argument?

Reply:
I've fixed this.